vllm/v1/engine/output_processor.py

```diff
@@ -74,6 +85,94 @@ def from_new_request(
             log_stats=log_stats,
         )
 
+    def make_request_output(
+        self,
+        new_token_ids: List[int],
+        finish_reason: Optional[FinishReason],
+        stop_reason: Union[int, str, None],
+    ) -> Optional[RequestOutput]:
+
+        finished = finish_reason is not None
+        output_kind = self.output_kind
+
+        # In follow up, we will switch to invariant where EngineCore
+        # does not stream partial prefills.
+        if not finished and (self.is_prefilling
+                             or output_kind == RequestOutputKind.FINAL_ONLY):
+            # Only the final output is required in FINAL_ONLY mode.
+            return None
+
+        # Aggregate child completions when not streaming
+        aggregate_children = output_kind == RequestOutputKind.FINAL_ONLY
+        request_output: Optional[RequestOutput] = None
+
+        if aggregate_children and self.output_aggregator is not None:
+            request_output = self.output_aggregator
+            self.output_aggregator = None
+        else:
+            request_output = self._new_request_output(
+                self.parent_req.request_id, finished)
+
+        completion_output = self._new_completion_output(
+            new_token_ids, finish_reason, stop_reason)
+        request_output.outputs.append(completion_output)
+
+        if (aggregate_children
+                and len(request_output.outputs) != self.parent_req.n):
+            self.output_aggregator = request_output
+            return None
+
+        return request_output
+
+    def _new_request_output(
+        self,
+        request_id: str,
+        finished: bool,
+    ) -> RequestOutput:
+
+        delta = self.output_kind == RequestOutputKind.DELTA
+
+        if delta:
+            # Side effect: logprobs processor forgets prompt logprobs
+            prompt_logprobs = self.logprobs_processor.pop_prompt_logprobs()
+        else:
+            prompt_logprobs = self.logprobs_processor.prompt_logprobs
+
+        return RequestOutput(
+            request_id=request_id,
+            prompt=self.prompt,
+            prompt_token_ids=self.prompt_token_ids,
+            prompt_logprobs=prompt_logprobs,
+            finished=finished,
+        )
+
+    def _new_completion_output(
+        self,
+        token_ids: List[int],
+        finish_reason: Optional[FinishReason],
+        stop_reason: Union[int, str, None],
+    ) -> CompletionOutput:
+
+        finished = finish_reason is not None
+        delta = self.output_kind == RequestOutputKind.DELTA
+
+        text = self.detokenizer.get_next_output_text(finished, delta)
+        if not delta:
+            token_ids = self.detokenizer.output_token_ids
+
+        logprobs = self.logprobs_processor.logprobs
+        if delta and logprobs:
+            logprobs = logprobs[-len(token_ids):]
```

Comment:
suggested small simplification:
```suggestion
        logprobs = self.logprobs_processor.logprobs
        if not delta:
            token_ids = self.detokenizer.output_token_ids
        elif logprobs:
            logprobs = logprobs[-len(token_ids):]
```

Reply:
Yeah, this layout was deliberate though - first prepare the text/token_ids param using the detokenizer, second prepare the logprobs. Both are distinct steps that both happen to depend on delta mode

You could even imagine like:

```
  # Prepare text and token_ids, based on delta mode
  text, token_ids = self.detokenizer.get_next_output(finished, delta)

  # Prepare logprobs, based on delta mode
  logprobs = self.logprobs_processor.get_logprobs(delta, token_ids)
```

Would comments like that help (even without the new methods)?

Anyway ... not a big deal, but there was a reason for doing it this way

Reply:
Makes sense, was only minor suggestion but I follow your reasoning.